### PR TITLE
Exp rpc 198

### PR DESCRIPTION
**Pull request by BgmLover**

### CODE REVIEW

File: src/rpc/rpc_tools.cpp [hunk 1]
```diff
@@ -41,14 +43,12 @@ rpc::Transaction* ToRPCTx(const Transaction& tx) {
 
     auto rpc_output = rpc_tx->mutable_outputs();
     for (const auto& out : tx.GetOutputs()) {
-        rpc_output->AddAllocated(ToRPCOutput(out));
+	rpc_output->AddAllocated(ToRPCOutput(out));
```
Comment (qdddddd) [inline src/rpc/rpc_tools.cpp]: Format

File: src/peer/peer_manager.cpp [hunk 2]
```diff
@@ -227,6 +228,9 @@ void PeerManager::ProcessTransaction(const ConstTxPtr& tx, PeerPtr& peer) {
     }
     if (MEMPOOL->ReceiveTx(tx)) {
         RelayTransaction(tx, peer);
+	if (PUBLISHER) {
+            PUBLISHER->PushMsg((void*)tx.get(), SubType::TX);
+	}
```
Comment (bassland) [inline src/peer/peer_manager.cpp]: The indent of these lines is tab. I suggest use 4 spaces.

File: src/rpc/subscription.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+
+class Publisher {
+public:
+    void AddNewSubscriber(std::string address, uint8_t service);
+    void DeleteSubscriber(std::string address);
+    void PushMsg(void* msg, SubType type);
+
+private:
+    std::unordered_map<std::string, Subscriber> subscribers_;
```
Comment (bassland) [inline src/rpc/subscription.h]: I think `subscribers_` is accessed in multi-thread, so I suggest add mutex lock before access the variable.

File: src/rpc/subscription.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
+#ifndef EPIC_SUBSCRIPTION_H
```
Comment (0ersted) [inline src/rpc/subscription.h]: credit heading

File: src/rpc/subscription.cpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+void Publisher::DeleteSubscriber(std::string address) {
+    subscribers_.erase(address);
+}
+
+void Publisher::PushMsg(void* msg, SubType type) {
+    switch (type) {
+        case SubType::BLOCK: {
+            rpc::Vertex vtx = *ToRPCVertex(*(Vertex*) msg);
```
Comment (0ersted) [inline src/rpc/subscription.cpp]: I think resources here are not deleted. Refer to rpc_tool.cpp. The same for line 49 `rpc::transaction`

File: src/rpc/subscription.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2019 EPI-ONE Core Developers
```
Comment (0ersted) [inline src/rpc/subscription.h]:
```suggestion
// Copyright (c) 2020 EPI-ONE Core Developers
```

File: src/consensus/chain.cpp [hunk 7]
```diff
@@ -261,6 +262,9 @@ VertexPtr Chain::Verify(const ConstBlockPtr& pblock) {
             vtx->UpdateReward(GetPrevReward(*vtx));
         }
         verifying_.insert({vtx->cblock->GetHash(), vtx});
+        if (PUBLISHER) {
+            PUBLISHER->PushMsg(vtx.get(), SubType::TX);
```
Comment (bassland) [inline src/consensus/chain.cpp]:
```suggestion
            PUBLISHER->PushMsg(vtx.get(), SubType::BLOCK);
```

**Pull request merged**